common/upload.js

```diff
@@ -0,0 +1,533 @@
+            var startUpload = function() {
+                vm.title = "Uploading files";
+                vm.isUpload = true;
+                vm.humanTotalSize = UiUtils.humanFileSize(vm.totalSize);
+
+                vm.queue = [];
+
+                vm.rows.forEach(function(row) {
+                    row.forEach(function(item) {
+                        item.onProgressChanged(0);
+                        vm.queue.push(item);
+                    });
+                });
+
+                startQueuedUpload();
+
+                speedIntervalTimer = setInterval(function() {
+                    var diff = vm.sizeTransferred - lastByteTransferred;
+                    lastByteTransferred = vm.sizeTransferred;
+
+                    if (diff <=0) vm.speed = 0;
+                    else vm.speed = UiUtils.humanFileSize(diff) + "ps";
+                }, 1000);
+
+            };
+
+            var startQueuedUpload = function() {
+                var item = vm.queue.shift();
+                if(!item) return;
+
+                item.hatracObj.start().then(
+                            item.onUploadCompleted.bind(item),
+                            onError, 
+                            item.onProgressChanged.bind(item));
+            };
+
+            // This function completes upload job in hatrac.js for all files
+            var completeUpload = function() {
+                vm.title = "Finalizing Upload";
+                vm.rows.forEach(function(row) {
+                    row.forEach(function(item) {
+                        item.hatracObj.completeUpload().then(
+                            item.onCompleteUploadJob.bind(item),
+                            onError);
+                    });
+                });
+            };
+
+            // This function aborts upload for all files
+            var abortUploads = function() {
+                clearInterval(speedIntervalTimer);
+                vm.aborted = true;
+                vm.speed = 0;
+                params.rows.forEach(function(row) {
+                    for(var k in row) {
+                        if (row[k] != null && typeof row[k] == 'object' && row[k].file) {
+                            row[k].hatracObj.cancel();
+                        }
+                    }
+                });
+            };
+
+            // This function is called by all rejected promises form above functions
+            var onError = function(err) {
+                if (vm.erred || vm.aborted) return;
+                
+                vm.erred = true;
+                
+                abortUploads();
+
+                $uibModalInstance.dismiss(err);
+            };
+
+            
+            /**
+             * @function
+             * @param {Object} col - Column Object
+             * @desc
+             * Creates an uploadFile obj to keep track of file and its upload.
+             */
+            var uploadFile = function(col, column, row) {
```

Comment:
I'm curious what the reasoning is for creating this object in `chaise` instead of it being in `ermrestJS`?

While this works, this should either be it's own angular `factory` or be defined in `ermrestJS`.